JAMES-2255 getMessagesList.position is long instead of int type

quynhn opened the pull request.

chibenwa commented on server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessageListMethodTest.java:
```diff
@@ -1810,4 +1811,30 @@ public void getMessageListShouldSortUsingInternalDateWhenNoDateHeader() throws E
+    public void getMessageListShouldAcceptAVeryBigNumberForPosition() throws Exception {
+        given()
+            .header("Authorization", aliceAccessToken.serialize())
+            .body("[[\"getMessageList\", {\"position\":9007199254740991}, \"#0\"]]")
```
Specs says "store until 2^53", I'd like this value to be explicit in the test

```
Math.pow(2, 53)
```

chibenwa commented on server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessageListMethodTest.java:
```diff
@@ -1810,4 +1811,30 @@ public void getMessageListShouldSortUsingInternalDateWhenNoDateHeader() throws E
+    }
+
+    @Test
+    public void getMessageListShouldErrorWhenPositionOutOfLongValue() throws Exception {
```
Things over 2^53 SHOULD NOT be supported.

chibenwa commented on server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessageListMethodTest.java:
```diff
@@ -1810,4 +1811,30 @@ public void getMessageListShouldSortUsingInternalDateWhenNoDateHeader() throws E
+            .statusCode(200)
+            .body(NAME, equalTo("error"))
+            .body(ARGUMENTS + ".type", equalTo("invalidArguments"))
+            .body(ARGUMENTS + ".description", containsString("out of range of long"));
```
The error message could be more specific:

```
Too big integer. All numbers used in this API are unsigned integers <= 2^53
```

chibenwa commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/GetMessageListRequest.java:
```diff
@@ -115,13 +115,19 @@ public Builder fetchSearchSnippets(boolean fetchSearchSnippets) {
         }
 
+        private void checkPosition() {
+            if (position != null) {
```
Then position should be an optional.

chibenwa commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/GetMessageListRequest.java:
```diff
@@ -115,13 +115,19 @@ public Builder fetchSearchSnippets(boolean fetchSearchSnippets) {
 
+        private void checkPosition() {
+            if (position != null) {
+                Preconditions.checkState(position >= 0, "'position' should be positive or null");
```
We may want to check here we are not higher than 2^53

chibenwa replied:
We need some strong typing here...

It would be wonderful if we serialize / deserialize every number into a `Number` that handles these checks for us... There might be some Jackson checks

```
Number.fromInt(a);
Number.fromLong(b);
number.getValue();
```

@JsonValue will probably be useful.

We need to use Number for all numeric values server by JMAP:

 - SortOrder
 - Mailboxes numeric properties
 - Messages numeric properties
 - etc...

chibenwa replied:
I sill think we need strong typing for Number. Could you please introduce a JMAP Number class that handles the knowledge of what a JMAP spec `Number` is ?

quynhn replied:
I don't think so, basically Integer and Long are extended from Number class and they had their limitation and I think it's enough, we do not need more Number class.

chibenwa replied:
No, I'm not speaking about the JDK Number class, I'm speaking about a specific **JMAP level** `Number` class, wrapping the definition of `what is a JMAP Number`. 

I don't see any downside argument about not doing it. I can help on this with some pair programming.

chibenwa commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/GetMessageListRequest.java:
```diff
@@ -45,16 +45,17 @@ public static Builder builder() {
         private final ImmutableList.Builder<String> fetchMessageProperties;
         private Boolean fetchSearchSnippets;
-
+private Number
```
Ident?

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/GetMessageListRequest.java:
```diff
@@ -115,19 +116,14 @@ public Builder fetchSearchSnippets(boolean fetchSearchSnippets) {
 
         public GetMessageListRequest build() {
-            checkPosition();
+            Preconditions.checkState(position.map(position -> position >= 0 && position <= Math.pow(2, 53)).orElse(true),
```
Math.pow(2, 53) should at least be a constant

chibenwa commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,58 @@
+        return new Number();
+    }
+
+    private long value;
```
This value should be final.

fromInt and fromLong should be static and you can delete factory()

chibenwa commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,58 @@
+
+    public long getValue() {
+        return value;
+    }
```
equals + hashcode

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,58 @@
+
+    private long value;
+    public Number fromInt(int value) {
+        System.out.println(value + " ||| " + MAX_INT_VALUE + " ||| " + (value >= MIN_VALUE && value <= MAX_INT_VALUE));
```
sysout spotted

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageFactoryTest.java:
```diff
@@ -24,6 +24,7 @@
 import static org.mockito.Mockito.when;
 
 import java.io.ByteArrayInputStream;
+import java.lang.*;
```
Wildcard spotted

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageFactoryTest.java:
```diff
@@ -83,7 +84,7 @@ public void emptyMailShouldBeLoadedIntoMessage() throws Exception {
         assertThat(testee)
             .extracting(Message::getPreview, Message::getSize, Message::getSubject, Message::getHeaders, Message::getDate)
-            .containsExactly("(Empty)", 0L, "", ImmutableMap.of("MIME-Version", "1.0"), INTERNAL_DATE);
+            .containsExactly("(Empty)", Number.lenientFactory().fromLong(0L), "", ImmutableMap.of("MIME-Version", "1.0"), INTERNAL_DATE);
```
ZERO?

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMailboxesMethodTest.java:
```diff
@@ -282,7 +282,7 @@ public void getMailboxesShouldReturnEmptyMailboxByDefault() throws MailboxExcept
                 .flatExtracting(GetMailboxesResponse::getList)
                 .extracting(Mailbox::getTotalMessages, Mailbox::getUnreadMessages)
-                .containsOnly(Tuple.tuple(0L, 0L));
+                .containsOnly(Tuple.tuple(Number.lenientFactory().fromLong(0L), Number.lenientFactory().fromLong(0L)));
```
ZERO?

chibenwa commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,133 @@
+
+    @JsonIgnore
+    public int asInt() {
+        return (int) value;
```
Ints.checkedCast(l)

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/AttachmentTest.java:
```diff
@@ -53,7 +53,7 @@ public void buildShouldThrowWhenTypeIsEmpty() {
     @Test
     public void buildShouldWorkWhenMandatoryFieldsArePresent() {
-        Attachment expected = new Attachment(BlobId.of("blobId"), "type", Optional.empty(), 123, Optional.empty(), false, Optional.empty(), Optional.empty());
+        Attachment expected = new Attachment(BlobId.of("blobId"), "type", Optional.empty(), Number.strictFactory().fromLong(123), Optional.empty(), false, Optional.empty(), Optional.empty());
```
extract numbers and give them a name?

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/AttachmentTest.java:
```diff
@@ -64,7 +64,8 @@ public void buildShouldWorkWhenMandatoryFieldsArePresent() {
     @Test
     public void buildShouldWorkWithAllFieldsSet() {
-        Attachment expected = new Attachment(BlobId.of("blobId"), "type", Optional.of("name"), 123, Optional.of("cid"), true, Optional.of(456L), Optional.of(789L));
+        Attachment expected = new Attachment(BlobId.of("blobId"), "type", Optional.of("name"), Number.strictFactory().fromLong(123), Optional.of("cid"), true,
```
Idem?

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/GetMailboxMessageListResponseTest.java:
```diff
@@ -75,7 +75,8 @@ public void builderShouldWork() {
         List<MessageId> messageIds = ImmutableList.of(TestMessageId.of(3), TestMessageId.of(4));
-        GetMessageListResponse expectedGetMessageListResponse = new GetMessageListResponse(null, filterCondition, sort, false, null, false, 0, 0, ImmutableList.of(), messageIds);
+        GetMessageListResponse expectedGetMessageListResponse = new GetMessageListResponse(null, filterCondition, sort, false, null, false,
+            Number.lenientFactory().fromInt(0), Number.lenientFactory().fromInt(0), ImmutableList.of(), messageIds);
```
ZERO

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageTest.java:
```diff
@@ -102,7 +102,7 @@ public void buildShouldThrowWhenPreviewIsEmpty() {
         Instant currentDate = Instant.now();
         Message expected = new Message(TestMessageId.of(1), BlobId.of("blobId"), "threadId", ImmutableList.of(InMemoryId.of(456)), Optional.empty(), false, ImmutableMap.of("key", "value"), Optional.empty(),
-                ImmutableList.of(), ImmutableList.of(), ImmutableList.of(), ImmutableList.of(), "subject", currentDate, 123, "preview", Optional.empty(), Optional.empty(),
+                ImmutableList.of(), ImmutableList.of(), ImmutableList.of(), ImmutableList.of(), "subject", currentDate, Number.strictFactory().fromLong(123), "preview", Optional.empty(), Optional.empty(),
```
Give that number a name?

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,87 @@
+public class NumberTest {
+    @Test
+    public void shouldRespectBeanContract() {
+        EqualsVerifier.forClass(Keywords.class).verify();
```
You should do that for number ;-)

quynhn replied:
I removed it

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,87 @@
+    @Test
+    public void fromLongShouldThrowWhenOver2Pow53Value() throws Exception {
+        assertThatThrownBy(() ->
+            Number.strictFactory().fromLong((long)Math.pow(2, 54)))
```
LongMath.pow ... Or use the constant intoduced in **Number**

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,87 @@
+    @Test
+    public void ensureLessThanThrowWhenOverSpecifiedValue() throws Exception {
+        assertThatThrownBy(() ->
+            Number.ensureLessThan(11, 10))
```
Number.factory().fromInt(11).ensureLessThat(10) seems more friendly...

chibenwa replied:
What happens when it is not the case? When they are equal?

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java:
```diff
@@ -63,7 +64,9 @@ public void buildShouldThrowWhenNameIsEmpty() {
     public void buildShouldWork() {
         Mailbox expectedMailbox = new Mailbox(InMemoryId.of(1), "name", Optional.of(InMemoryId.of(0)), Optional.of(Role.DRAFTS), SortOrder.of(123),
-                true, true, true, true, true, true, true, 456, 789, 741, 852, Rights.EMPTY, MailboxNamespace.personal());
+            true, true, true, true, true, true, true,
```
Extract these numbers?

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java:
```diff
@@ -105,7 +108,7 @@ public void totalMessagesShouldNeverBeNegative() {
                 .build();
 
-        assertThat(mailbox.getTotalMessages()).isEqualTo(0);
+        assertThat(mailbox.getTotalMessages().asLong()).isEqualTo(0);
```
Asserts on Number?

assertThat(mailbox.getTotalMessages())
   .isEqualTo(Number.lenientFactory().fromInt(0)));

chibenwa commented on server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessageListMethodTest.java:
```diff
@@ -1898,4 +1899,30 @@ public void getMessageListShouldSortUsingInternalDateWhenNoDateHeader() throws E
+    public void getMessageListShouldAcceptLessThan2Pow53NumberForPosition() throws Exception {
+        given()
+            .header("Authorization", aliceAccessToken.serialize())
+            .body("[[\"getMessageList\", {\"position\":" + Math.pow(2, 53) + "}, \"#0\"]]")
```
Can't we use Number.MAX_LONG_VALUE ?

chibenwa commented on server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessageListMethodTest.java:
```diff
@@ -1898,4 +1899,30 @@ public void getMessageListShouldSortUsingInternalDateWhenNoDateHeader() throws E
+    public void getMessageListShouldErrorWhenPositionOver2Pow53() throws Exception {
+        given()
+            .header("Authorization", aliceAccessToken.serialize())
+            .body("[[\"getMessageList\", {\"position\":" + Math.pow(2, 54) + "}, \"#0\"]]")
```
Number.MAX_LONG_VALUE + 1

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,97 @@
+    @Test
+    public void fromIntShouldReturnMinValueWhenNegativeValueWithLenient() throws Exception {
+        assertThat(Number.lenientFactory().fromInt(-1).asInt())
+            .isEqualTo(0);
```
ZERO

chibenwa commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,97 @@
+    @Test
+    public void fromLongShouldReturnMinValueWhenNegativeValueWithLenient() throws Exception {
+        assertThat(Number.lenientFactory().fromLong(-1).asLong())
+            .isEqualTo(0);
```
Idem

rouazana commented on server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessageListMethodTest.java:
```diff
@@ -1898,4 +1900,30 @@ public void getMessageListShouldSortUsingInternalDateWhenNoDateHeader() throws E
+    public void getMessageListShouldErrorWhenPositionOver2Pow53() throws Exception {
+        given()
+            .header("Authorization", aliceAccessToken.serialize())
+            .body("[[\"getMessageList\", {\"position\":" + Number.MAX_LONG_VALUE + 1 + "}, \"#0\"]]")
```
why MAX_LONG_VALUE? that's MAX_VALUE no? Long.MAX_VALUE is different

chibenwa replied:
According to JMAP spec (http://jmap.io/spec-core.html#the-number-datatype) a number can not be greater than 2^53, this is that max value.

Number here being a JMAP object, I feel it's pretty clear that max value is the value from the spec.

rouazana replied:
yes, so that's what I'm saying, it should be MAX_VALUE not MAX_LONG_VALUE.

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessageListMethod.java:
```diff
@@ -125,9 +127,9 @@ private GetMessageListResponse getMessageListResponse(GetMessageListRequest mess
+                messageListRequest.getLimit().map(Number::asInt).orElse(maximumLimit) + messageListRequest.getPosition().map(Number::asLong).orElse(DEFAULT_POSITION))
                 .stream()
-                .skip(messageListRequest.getPosition())
+                .skip(messageListRequest.getPosition().map(Number::asLong).orElse(DEFAULT_POSITION))
```
why asLong here but asInt just before?

quynhn replied:
I know it does not consistent but the limit is int value and position is long value
Anyway, the parameter of function is long so I changed them all to asLong

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java:
```diff
@@ -163,7 +163,7 @@ public Builder date(Instant date) {
 
         public Builder size(long size) {
-            this.size = size;
+            this.size = Number.lenientFactory().fromLong(size);
```
why lenient here?

chibenwa replied:
It's from our code -> to the user. As far as I know we never parse Messages (but CreationMessage).

If we don't want to fail badly on negative size (for instance) we have to use a lenient stuff.

rouazana replied:
then Number.fromOutboundLong(size) or something like that?

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,134 @@
+    public static class StrictFactory implements Factory {
+        public Number fromInt(int value) {
+            Preconditions.checkState(value >= ZERO_VALUE,
+                "value should be positive and less than 2^31 or empty");
```
or empty?

quynhn replied:
This error message should display to the end-user so it makes sense to tell them that value should be restrict or let it be empty. WDYT?

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,134 @@
+
+        public Number fromLong(long value) {
+            Preconditions.checkState(value >= ZERO_VALUE && value <= MAX_LONG_VALUE,
+                "value should be positive and less than 2^53 or empty");
```
or empty?

quynhn replied:
idem

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    public static Factory lenientFactory() {
```
do we really need this?

chibenwa replied:
I guess so.

Every time we return a Number value we need it to fit the **Number** format. Take the recent `unread count can be negative issue`. Do you think that is the only affected property? By design here we will be safe against this (and with logs ;-) )

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,134 @@
+
+    public Number ensureLessThan(long maxValue) {
+        Preconditions.checkState(value >= ZERO_VALUE && value <= maxValue,
+            "value should be positive and less than " + maxValue + " or empty");
```
again this "or empty"?

quynhn replied:
Same above anwser

chibenwa replied:
I will modify this error message.

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Mailbox.java:
```diff
@@ -136,22 +139,22 @@ public Builder mayDelete(boolean mayDelete) {
 
         public Builder totalMessages(long totalMessages) {
-            this.totalMessages = Optional.of(totalMessages);
+            this.totalMessages = Optional.of(Number.lenientFactory().fromLong(totalMessages));
```
how do you decide sometimes lenient, sometimes strict?

chibenwa replied:
When it's parsed from the user input -> strict

When it's coming from our code -> lenient

rouazana replied:
strange that's the contrary of the motto: be strict with what you send, be tolerant with what you get (my free interpretation)

chibenwa replied:
We can very well decide to use lenient stuff on what we receive as well. But I don't like modifying user input in the explicit protocol that JMAP is supposed to be.

Now, if we want to be strict about what we send:
 - it means we won't display ie some mailboxes when their count is negative. Or we might have similar problems with similar values.
 - we anyway need to push the sanitizing lower in the stack, IMO to the Dao

rouazana commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/FilterToSearchQuery.java:
```diff
@@ -77,8 +77,8 @@ private SearchQuery convertCondition(FilterCondition filter) {
-        filter.getMaxSize().ifPresent(maxSize -> searchQuery.andCriteria(SearchQuery.sizeLessThan(maxSize)));
-        filter.getMinSize().ifPresent(minSize -> searchQuery.andCriteria(SearchQuery.sizeGreaterThan(minSize)));
+        filter.getMaxSize().ifPresent(maxSize -> searchQuery.andCriteria(SearchQuery.sizeLessThan(maxSize.asInt())));
+        filter.getMinSize().ifPresent(minSize -> searchQuery.andCriteria(SearchQuery.sizeGreaterThan(minSize.asInt())));
```
why ints here?

quynhn replied:
Changed to asLong because the function's parameter is long value

rouazana commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,97 @@
+    @Test
+    public void fromLongShouldReturnMinValueWhenNegativeValueWithLenient() throws Exception {
+        assertThat(Number.lenientFactory().fromLong(-1))
+            .isEqualTo(Number.ZERO);
```
at this point this is not lenient, this is normalization

chibenwa replied:
Doesn't all lenient parser apply normalization to not normal input in order to get a normal output?

I mean the normalization you apply in a lenient parser can be ignoring values/fields, sanitizing values, etc...

rouazana commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,97 @@
+    public void ensureLessThanShouldThrowWhenOverSpecifiedValue() throws Exception {
+        assertThatThrownBy(() ->
+            Number.strictFactory().fromInt(11).ensureLessThan(10))
+            .isInstanceOf(IllegalStateException.class);
```
indent

quynhn commented:
> Not sure to understand the desing.
Also I think you are missing an integration test where you submit a number > Integer.MAX_VALUE which will be transformed as int.

In fact I added but instead of throwing exception, it's accept with Integer.MAX_VALUE so I removed it. For the value has to < Integer.MAX_VALUE, we will use "ensureLessThan".

aduprat commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Test
+    public void fromLongShouldThrowWhenOver2Pow53Value() throws Exception {
```
You may add such test for integers _fromInt_.
The error message is saying that there's a max value, but it's not coded.

chibenwa commented:
```
[06fbc25e5347aa1dee2672ca05656ed12013180d] [ERROR]   Java heap space
```

chibenwa commented:
test this please

chibenwa commented:
```
[acf91dc1b105aa179c39b59b61368f42a63c0233] [ERROR]   Java heap space
```

test this please

rouazana commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/model/NumberTest.java:
```diff
@@ -0,0 +1,66 @@
+
+public class NumberTest {
+    @Test
+    public void fromLongShouldReturnMinValueWhenNegativeValueWithLenient() throws Exception {
```
Lenient does not exist any more

chibenwa commented:
test this please

chibenwa commented:
test this please

chibenwa commented:
test this please

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,90 @@
+    public static final Logger LOGGER = LoggerFactory.getLogger(Number.class);
+
+    public static Number fromLong(long value) {
+        Preconditions.checkState(value >= ZERO_VALUE && value <= MAX_VALUE,
```
I would not throw but return an Optional (with don't have Either yet).
That way, one can choose what to do with the incorrect value using `orElse`

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Number.java:
```diff
@@ -0,0 +1,90 @@
+        return new Number(value);
+    }
+
+    public static Number fromOutboundLong(long value) {
```
this method name don't talk to me. And I have no idea it would actually do a MAX or MIN when out of bound

chibenwa replied:
I called it BOUND_SANITIZING_FACTORY

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Mailbox.java:
```diff
@@ -170,12 +173,7 @@ public Mailbox build() {
-        private long negativeToZero(Optional<Long> number) {
-            return number.filter(value -> value >= 0)
-                .orElse(0L);
+                    totalMessages.orElse(Number.ZERO), unreadMessages.orElse(Number.ZERO), totalThreads.orElse(Number.ZERO), unreadThreads.orElse(Number.ZERO), sharedWith.orElse(Rights.EMPTY), namespace.orElse(MailboxNamespace.personal()));
```
please wrap this line

chibenwa commented:
Merged
